playwright/errors.spec.ts

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2025 New Vector Ltd.
+
+SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
+Please see LICENSE in the repository root for full details.
+*/
+
+import { expect, test } from "@playwright/test";
+
+test("Should show error screen if fails to get JWT token", async ({ page }) => {
+  await page.goto("/");
+
+  await page.getByTestId("home_callName").click();
+  await page.getByTestId("home_callName").fill("HelloCall");
+  await page.getByTestId("home_displayName").click();
+  await page.getByTestId("home_displayName").fill("John Doe");
+  await page.getByTestId("home_go").click();
+
+  await page.route("**/openid/request_token", (route) =>
+    route.fulfill({
+      // 418 is a non retryable error, so test will fail immediately
+      status: 418,
+    }),
+  );
+
+  // Join the call
+  await page.getByTestId("lobby_joinCall").click();
+
+  // Should fail
+  await expect(page.getByText("Something went wrong")).toBeVisible();
+  await expect(page.getByText("OPEN_ID_ERROR")).toBeVisible();
+});
+
+test("Should automatically retry non fatal JWT errors", async ({ page }) => {
+  await page.goto("/");
+
+  await page.getByTestId("home_callName").click();
+  await page.getByTestId("home_callName").fill("HelloCall");
+  await page.getByTestId("home_displayName").click();
+  await page.getByTestId("home_displayName").fill("John Doe");
+  await page.getByTestId("home_go").click();
+
+  let firstCall = true;
+  let hasRetriedCallback: (value: PromiseLike<void> | void) => void;
+  let hasRetriedPromise = new Promise<void>((resolve) => {
+    hasRetriedCallback = resolve;
+  });
+  await page.route("**/openid/request_token", (route) => {
+    if (firstCall) {
+      firstCall = false;
+      route.fulfill({
+        status: 429,
+      });
+    } else {
+      route.continue();
+      hasRetriedCallback();
+    }
+  });
+
+  // Join the call
+  await page.getByTestId("lobby_joinCall").click();
+  // Expect that the call has been retried
+  await hasRetriedPromise;
+  await expect(page.getByText("Something went wrong")).not.toBeVisible();
```

Comment:
I'm wary of async assertions that expect something to *not* exist - it seems there is opportunity for the assertion to succeed before the page actually updates to the state it's meant to be testing. Is there an equivalent, positive assertion that this could be replaced with?

Reply:
Agree, I wanted to do something like that but I didn't see anything obvious in the UI showing that I have or not the token.

But maybe I could create a fixture where the tester is invited to an existing call, and then I could check that there are 2 participants?

Reply:
by adding an inviter and multiple page to the test it was malking it a bit flacky. So I just tested that the API is called a second time and then that the video element is displayed